File: basic stuff/3sum.js
// 3 sum, given an array of numbers, return all triplets that sum to 0;
// Given array nums = [-1, 0, 1, 2, -1, -4],
// [
//   [-1, 0, 1],
//   [-1, -1, 2]
// ]
const threeSum = (nums, target=0) => {
  const result = [];

  if (nums.length < 3) return result;

  let sorted = nums.sort((el1, el2) => {
    if (el1 > el2) return 1;
    if (el2 > el1) return -1;
    return 0
  });

  for (let i = 0; i < sorted.length - 2; i++) {
    if (i === 0 || sorted[i] > sorted[i - 1]) {
      let start = i + 1;
      let end = sorted.length - 1;
      while (start < end) {
        if (sorted[start] + sorted[end] + sorted[i] === target) {
          result.push([sorted[i], sorted[start], sorted[end]]);
        }
  
        if (sorted[start] + sorted[end] + sorted[i] < target) {
          let curr_start = start;
          while (sorted[start] === sorted[curr_start]  && start < end) {
            start += 1;
          }
  
        } else {
          let curr_end = end;
          while (sorted[end] === sorted[curr_end] && start < end) {
            end -= 1;
          }
  
        }
      }
    }
  }
  return result;
}

console.log(threeSum([-1,0,1,2,-1,-4]));

File: basic stuff/leetcode/3sum.py
class Solution(object):
    def threeSum(self, nums):
        res = []
        nums.sort()
        for i in range(len(nums) - 2):
            if (i == 0 or nums[i] > nums[i - 1]):
                start = i + 1
                end = len(nums) - 1
                while start < end:
                    if nums[start] + nums[end] + nums[i] == 0:
                        res.append([nums[start], nums[end], nums[i]])
                    if nums[start] + nums[end] + nums[i] < 0:
                        curr_start = start
                        while (nums[start] == nums[curr_start] and start < end):
                            start += 1
                    else:
                        curr_end = end
                        while (nums[end] == nums[curr_end] and end > start):
                            end -= 1
                            
        return res


res = Solution()

print(res.threeSum([-1, -1, 0, 1, 2, -4]))